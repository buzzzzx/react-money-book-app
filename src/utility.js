export const LIST_VIEW = "list";
export const CHART_VIEW = "chart";
