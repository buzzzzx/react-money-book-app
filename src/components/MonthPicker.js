import React, { Component, createRef } from "react";
import PropTypes from "prop-types";
import { padLeft, range } from "../utility";

class MonthPicker extends Component {
  state = {
    isOpen: false,
    selectedYear: this.props.year,
    selectedMonth: this.props.month,
  };

  dropdownRef = createRef();

  componentDidMount() {
    document.addEventListener("click", this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener("click", this.handleClickOutside);
  }

  handleClickOutside = (event) => {
    if (this.dropdownRef && !this.dropdownRef.current.contains(event.target)) {
      // means click outside
      this.setState({
        isOpen: false,
      });
    }
  };

  toggleDropdown = (event) => {
    event.preventDefault();
    this.setState({
      isOpen: !this.state.isOpen,
    });
  };

  selectYear = (event, year) => {
    event.preventDefault();
    this.setState({
      selectedYear: year,
    });
  };

  selectMonth = (event, month) => {
    event.preventDefault();
    this.setState({
      selectedMonth: month,
      isOpen: false,
    });
    this.props.onChange(this.state.selectedYear, month);
  };

  generateActiveClass = (current, target) => {
    return current === target ? "dropdown-item active" : "dropdown-item";
  };

  render() {
    const { isOpen, selectedMonth, selectedYear } = this.state;
    const yearRange = range(9, -4).map((num) => num + 2020);
    const monthRange = range(12, 1);

    return (
      <div
        className="dropdown my-4 month-picker-component"
        ref={this.dropdownRef}
      >
        <span className="mr-2">选择时间</span>
        <button
          className="btn btn-lg btn-secondary dropdown-toggle"
          onClick={this.toggleDropdown}
        >
          {`${selectedYear} 年 ${padLeft(selectedMonth)} 月`}
        </button>
        {isOpen && (
          <div className="dropdown-menu" style={{ display: "block" }}>
            <div className="row">
              <div className="col border-right years-range">
                {yearRange.map((yearNum, index) => (
                  <a
                    href="#"
                    className={this.generateActiveClass(yearNum, selectedYear)}
                    key={index}
                    onClick={(event) => this.selectYear(event, yearNum)}
                  >
                    {yearNum} 年
                  </a>
                ))}
              </div>
              <div className="col months-range">
                {monthRange.map((monthNum, index) => (
                  <a
                    href="#"
                    className={this.generateActiveClass(
                      monthNum,
                      selectedMonth
                    )}
                    key={index}
                    onClick={(event) => this.selectMonth(event, monthNum)}
                  >
                    {padLeft(monthNum)} 月
                  </a>
                ))}
              </div>
            </div>
          </div>
        )}
      </div>
    );
  }
}

MonthPicker.propTypes = {
  year: PropTypes.number.isRequired,
  month: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default MonthPicker;
