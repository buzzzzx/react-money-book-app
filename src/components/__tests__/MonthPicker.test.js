import React from "react";
import { mount } from "enzyme";
import { findDOMNode } from "react-dom";

import { padLeft } from "../../utility";
import MonthPicker from "../MonthPicker";

const props = {
  year: 2020,
  month: 6,
  onChange: jest.fn(),
};

let wrapper;

describe("test <MonthPicker />", () => {
  beforeEach(() => {
    wrapper = mount(<MonthPicker {...props} />);
  });

  it("should render the component to match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render the correct result by default props", () => {
    const dateText = wrapper.find(".dropdown-toggle").text();
    expect(dateText).toEqual("2020 年 06 月");

    expect(wrapper.find(".dropdown-menu").length).toEqual(0);
    expect(wrapper.state("isOpen")).toEqual(false);
  });

  it("after click dropdown button, should render year&month lists", () => {
    wrapper.find(".dropdown-toggle").simulate("click");
    expect(wrapper.find(".dropdown-menu").length).toEqual(1);
    expect(wrapper.state("isOpen")).toEqual(true);

    // check active year&month
    expect(wrapper.find(".years-range .dropdown-item.active").text()).toEqual(
      `${props.year} 年`
    );
    expect(wrapper.find(".months-range .dropdown-item.active").text()).toEqual(
      `${padLeft(props.month)} 月`
    );

    // check the length of year&month lists
    expect(wrapper.find(".years-range .dropdown-item").length).toEqual(9);
    expect(wrapper.find(".months-range .dropdown-item").length).toEqual(12);

    // check first year should be current year minus 4, and first month should be 01
    expect(wrapper.find(".years-range .dropdown-item").first().text()).toEqual(
      `${props.year - 4} 年`
    );
    expect(wrapper.find(".months-range .dropdown-item").first().text()).toEqual(
      "01 月"
    );
  });

  it("after click year&month item, should trigger the right status change", () => {
    wrapper.find(".dropdown-toggle").simulate("click");

    wrapper.find(".years-range .dropdown-item").first().simulate("click");
    expect(wrapper.state("selectedYear")).toEqual(props.year - 4);
    expect(
      wrapper.find(".years-range .dropdown-item").first().hasClass("active")
    ).toEqual(true);

    const firstMonthItem = wrapper.find(".months-range .dropdown-item").first();
    firstMonthItem.simulate("click");
    expect(wrapper.state("selectedMonth")).toEqual(1);
    expect(wrapper.state("isOpen")).toEqual(false);
    expect(props.onChange).toHaveBeenCalledWith(props.year - 4, 1);
  });

  it("when dropdown is showing, click the outside, dropdown should be closed", () => {
    const eventMap = {};
    // jest.fn() will return a new. unused function
    // jest.fn(implementation), implementation is a mock implementation of left hand side
    document.addEventListener = jest.fn((event, cb) => {
      eventMap[event] = cb;
    });

    // mock componentDidMout: immediately add event listener after component mount
    const wrapper = mount(<MonthPicker {...props} />);

    wrapper.find(".dropdown-toggle").simulate("click");
    eventMap.click({
      target: findDOMNode(wrapper.instance()),
    });
    expect(wrapper.state("isOpen")).toEqual(true);

    eventMap.click({
      target: document,
    });
    expect(wrapper.state("isOpen")).toEqual(false);
  });
});
