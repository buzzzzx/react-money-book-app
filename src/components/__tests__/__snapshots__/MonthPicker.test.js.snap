// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <MonthPicker /> should render the component to match the snapshot 1`] = `ReactWrapper {}`;
